core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java

```diff
@@ -155,101 +101,14 @@ public static UnderFileSystemFactory find(
-
-    List<UnderFileSystemFactory> factories = new ArrayList<>(FACTORIES);
-    scanLibs(factories);
-    scanExtensions(factories);
-
-    List<UnderFileSystemFactory> eligibleFactories = new ArrayList<>();
-    for (UnderFileSystemFactory factory : factories) {
-      if (factory.supportsPath(path, ufsConf)) {
-        LOG.debug("Under File System Factory implementation {} is eligible for path {}",
-            factory.getClass(), path);
-        eligibleFactories.add(factory);
-      }
-    }
-
-    if (eligibleFactories.isEmpty()) {
-      LOG.warn("No Under File System Factory implementation supports the path {}", path);
-    }
-    return eligibleFactories;
-  }
-
-  /**
-   * Finds all {@link UnderFileSystemFactory} from the extensions directory.
-   *
-   * @param factories list of factories to add to
-   */
-  private static void scanExtensions(List<UnderFileSystemFactory> factories) {
-    LOG.info("Loading extension UFS jars from {}", Configuration.get(PropertyKey.EXTENSIONS_DIR));
-    scan(Arrays.asList(ExtensionUtils.listExtensions()), factories);
-  }
-
-  /**
-   * Finds all {@link UnderFileSystemFactory} from the lib directory.
-   *
-   * @param factories list of factories to add to
-   */
-  private static void scanLibs(List<UnderFileSystemFactory> factories) {
-    LOG.info("Loading core UFS jars from {}", LIB_DIR);
-    List<File> files = new ArrayList<>();
-    try (DirectoryStream<Path> stream =
-        Files.newDirectoryStream(Paths.get(LIB_DIR), "alluxio-underfs-*.jar")) {
-      for (Path entry : stream) {
-        if (entry.toFile().isFile()) {
-          files.add(entry.toFile());
-        }
-      }
-    } catch (IOException e) {
-      LOG.warn("Failed to load UFS libs: {}", e.getMessage());
-    }
-    scan(files, factories);
-  }
-
-  /**
-   * Class-loads jar files that have not been loaded.
-   *
-   * @param files jar files to class-load
-   * @param factories list of factories to add to
-   */
-  private static void scan(List<File> files, List<UnderFileSystemFactory> factories) {
-    for (File jar : files) {
-      try {
-        URL extensionURL = jar.toURI().toURL();
-        String jarPath = extensionURL.toString();
-
-        ClassLoader extensionsClassLoader = new ExtensionsClassLoader(new URL[] {extensionURL},
-            ClassLoader.getSystemClassLoader());
-        ServiceLoader<UnderFileSystemFactory> extensionServiceLoader =
-            ServiceLoader.load(UnderFileSystemFactory.class, extensionsClassLoader);
-        for (UnderFileSystemFactory factory : extensionServiceLoader) {
-          LOG.debug("Discovered an Under File System Factory implementation {} - {} in jar {}",
-              factory.getClass(), factory.toString(), jarPath);
-          register(factory, factories);
-        }
-      } catch (Throwable t) {
-        LOG.warn("Failed to load jar {}: {}", jar, t.getMessage());
-      }
-    }
+    return sRegistryInstance.findAll(path, ufsConf);
```

Comment:
It seems safer to wrap `sRegistryInstance` in a getInstance method to protect against null references https://crunchify.com/thread-safe-and-a-fast-singleton-implementation-in-java/

Reply:
This sounds like a good idea.

Reply:
The static initializer of the class calls to the `init()` method to ensure the instance is created before anyone can access it. This is unchanged in this PR from how we originally initialize resources with static functions.